### PR TITLE
sanitize arrays when logging as hyperparameters in TensorBoardLogger

### LINKED ISSUE #9022: Logging tensors as hparam fails
## 🐛 Bug
When a trainer starts `fit()` and logging hyper parameters, an error occurs if one of hyperparameters is a `Tensor`.

### To Reproduce

```python
import pytorch_lightning as pl
import torch
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader


class Model(pl.LightningModule):
    def __init__(self, some_tensor, num):
        super().__init__()
        self.save_hyperparameters()  # Not OK
        # self.save_hyperparameters(ignore="some_tensor")  # OK
        self.tensor0 = torch.nn.Parameter(some_tensor, requires_grad=False)
        self.tensor1 = torch.nn.Parameter(torch.ones(1))
        self.num = num

    def forward(self, batch):
        return self.tensor0.sum() + batch + self.tensor1

    def training_step(self, batch, batch_idx):
        input, target = batch
        pred = self(input)
        return F.mse_loss(pred, target)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters())


class RandomData(Dataset):
    def __getitem__(self, item):
        return torch.randn(1)

    def __len__(self):
        return 100


if __name__ == "__main__":
    dataset = RandomData()
    pl.seed_everything(1, workers=True)
    tensor = torch.randn(2, 3)
    model = Model(tensor, 1)
    dataloader = DataLoader(dataset, batch_size=2, num_workers=2)
    trainer = pl.Trainer(gpus=[1, 2],
                         max_steps=10,
                         accelerator="ddp")
    trainer.fit(model, dataloader)

```

### Expected behavior
If I ignore the `some_tensor`, the trainer works fine, but if I don't, the trainer will give me an error and keep one of my GPU (GPU-2) running at 100%, which is very weird to me.

The error is 
```shell
Traceback (most recent call last):
  File "/home/liangf/research/SemanticNetMPL/temp/test.py", line 45, in <module>
    trainer.fit(model, dataloader)
  File "/home/liangf/miniconda3/envs/torch/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 553, in fit
    self._run(model)
  File "/home/liangf/miniconda3/envs/torch/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 912, in _run
    self._pre_dispatch()
  File "/home/liangf/miniconda3/envs/torch/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 941, in _pre_dispatch
    self._log_hyperparams()
  File "/home/liangf/miniconda3/envs/torch/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 968, in _log_hyperparams
    self.logger.log_hyperparams(hparams_initial)
  File "/home/liangf/miniconda3/envs/torch/lib/python3.9/site-packages/pytorch_lightning/utilities/distributed.py", line 48, in wrapped_fn
    return fn(*args, **kwargs)
  File "/home/liangf/miniconda3/envs/torch/lib/python3.9/site-packages/pytorch_lightning/loggers/tensorboard.py", line 196, in log_hyperparams
    exp, ssi, sei = hparams(params, metrics)
  File "/home/liangf/miniconda3/envs/torch/lib/python3.9/site-packages/torch/utils/tensorboard/summary.py", line 192, in hparams
    ssi.hparams[k].number_value = v
TypeError: array([0.66135216, 0.2669241 , 0.06167726], dtype=float32) has type numpy.ndarray, but expected one of: int, long, float
```
This seems to indicate that we cannot save tensors as hyperparameters, but interestingly, there are codes checking if a hyperparameter is a `Tensor` and converting it to numpy array, see this
```python
     if isinstance(v, torch.Tensor):
            v = make_np(v)[0]
            ssi.hparams[k].number_value = v  # torch/lib/python3.9/site-packages/torch/utils/tensorboard/summary.py
            hps.append(HParamInfo(name=k, type=DataType.Value("DATA_TYPE_FLOAT64")))
            continue
```
As for the running forever problem, I think it is because the process at rank 0 is died from this error, but the spawned process is not and it runs wild.

### Environment

* CUDA:
        - GPU:
                - NVIDIA RTX A6000
                - NVIDIA RTX A6000
                - NVIDIA RTX A6000
                - NVIDIA RTX A6000
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.20.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.9.0
        - pytorch-lightning: 1.4.1
        - tqdm:              4.61.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.9.2
        - version:           #90-Ubuntu SMP Fri Jul 9 22:49:44 UTC 2021

### Additional context

I think an old [issue](https://github.com/PyTorchLightning/pytorch-lightning/issues/4496) may also be caused by this problem, but no one tried to take a closer look at it.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #9022

TensorBoardLogger can log hyperparameters but not all types of data. It can do arrays that are 1-dimensional but anything higher-dimensional not, so we will sanitize it by converting to string as we do with other unsupported types.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/tensorboard.py [hunk 1]
```diff
@@ -286,6 +287,12 @@ def _get_next_version(self):
 
         return max(existing_versions) + 1
 
+    @staticmethod
+    def _sanitize_params(params: Dict[str, Any]) -> Dict[str, Any]:
+        params = LightningLoggerBase._sanitize_params(params)
+        # logging of arrays with dimension > 1 is not supported, sanitize as string
+        return {k: str(v) for k, v in params.items() if isinstance(v, (torch.Tensor, np.ndarray)) and v.ndim > 1}
```
Comment (justusschock) [inline pytorch_lightning/loggers/tensorboard.py:294]:
```suggestion
        return {k: str(v) if isinstance(v, (torch.Tensor, np.ndarray)) and v.ndim > 1 else v for k, v in params.items() }
```

We should keep all params that are no tensors or arrays as well I think :)

Comment (awaelchli) [reply pytorch_lightning/loggers/tensorboard.py:294]: absolutely!

**Review by tchaton — APPROVED** [review]
Neat !

File: pytorch_lightning/loggers/tensorboard.py [hunk 2]
```diff
@@ -286,6 +287,12 @@ def _get_next_version(self):
 
         return max(existing_versions) + 1
 
+    @staticmethod
+    def _sanitize_params(params: Dict[str, Any]) -> Dict[str, Any]:
+        params = LightningLoggerBase._sanitize_params(params)
+        # logging of arrays with dimension > 1 is not supported, sanitize as string
+        return {k: str(v) if isinstance(v, (torch.Tensor, np.ndarray)) and v.ndim > 1 else v for k, v in params.items()}
```
Comment (tchaton) [inline pytorch_lightning/loggers/tensorboard.py:294]: Should we add pd.dataFrame ?

Comment (awaelchli) [reply pytorch_lightning/loggers/tensorboard.py:294]: pandas is a dependency of Lightning, so I would say no.

File: pytorch_lightning/loggers/tensorboard.py [hunk 3]
```diff
@@ -286,6 +287,12 @@ def _get_next_version(self):
 
         return max(existing_versions) + 1
 
+    @staticmethod
+    def _sanitize_params(params: Dict[str, Any]) -> Dict[str, Any]:
+        params = LightningLoggerBase._sanitize_params(params)
```
Comment (ananthsub) [inline pytorch_lightning/loggers/tensorboard.py:292]: For later, should we move this out as a utility function? Are there opportunities for consolidation with other hyper parameter saving logic the framework has?

**Review by ananthsub — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/loggers/tensorboard.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/loggers/tensorboard.py:292]: We can but keep in mind not all loggers have the same level of support and flexibility so sanitization, flattening etc. is generally speaking always subject to the individual logger.

> Are there opportunities for consolidation with other hyper parameter saving logic the framework has?

not any I know of

**Pull request merged**